### PR TITLE
Added restriction on the black list of IP addresses

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by kruzhambus — APPROVED** [review]

**Review by Blaumaus — APPROVED** [review]

File: src/common/constants.ts [hunk 1]
```diff
@@ -131,6 +131,9 @@ const HISTORY_LIFE_TIME_DAYS = 30
 const PID_REGEX = /^(?!.*--)[a-zA-Z0-9-]{12}$/
 const isValidPID = (pid: string) => PID_REGEX.test(pid)
 
+const IP_REGEX = /^(((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(((\/([4-9]|[12][0-9]|3[0-2]))?)|\s?-\s?((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))))(,\s?|$))+/
```
Comment (Blaumaus) [inline src/common/constants.ts:134]: This `IP_REGEX` works for IPv4 (and IPv4 masks) only.
Could you add a new REGEX for IPv6 addresses too and use it in `isValidIP` function like `IPV4_REGEX.test(ip) || IPV6_REGEX.test(ip)` ?

**Review by Blaumaus — CHANGES_REQUESTED** [review]

**Review by kruzhambus — COMMENTED** [review]
maybe you can try this
const ipV4Regex = "(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)(?:\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)){3}";

const IpV6OctateRegex = "[a-fA-F\\d]{1,4}";

const ipV6Regex = `
(?:
(?:${IpV6OctateRegex}:){7}(?:${IpV6OctateRegex}|:)|                                   
(?:${IpV6OctateRegex}:){6}(?:${ipV4Regex}|:${IpV6OctateRegex}|:)|                             
(?:${IpV6OctateRegex}:){5}(?::${ipV4Regex}|(?::${IpV6OctateRegex}){1,2}|:)|                   
(?:${IpV6OctateRegex}:){4}(?:(?::${IpV6OctateRegex}){0,1}:${ipV4Regex}|(?::${IpV6OctateRegex}){1,3}|:)| 
(?:${IpV6OctateRegex}:){3}(?:(?::${IpV6OctateRegex}){0,2}:${ipV4Regex}|(?::${IpV6OctateRegex}){1,4}|:)| 
(?:${IpV6OctateRegex}:){2}(?:(?::${IpV6OctateRegex}){0,3}:${ipV4Regex}|(?::${IpV6OctateRegex}){1,5}|:)| 
(?:${IpV6OctateRegex}:){1}(?:(?::${IpV6OctateRegex}){0,4}:${ipV4Regex}|(?::${IpV6OctateRegex}){1,6}|:)| 
(?::(?:(?::${IpV6OctateRegex}){0,5}:${ipV4Regex}|(?::${IpV6OctateRegex}){1,7}|:))            
)(?:%[0-9a-zA-Z]{1,})?                                            
`
  .replace(/\s*\/\/.*$/gm, "")
  .replace(/\n/g, "")
  .trim()

File: src/project/project.service.ts [hunk 2]
```diff
@@ -219,6 +219,11 @@ export class ProjectService {
     if (_size(projectDTO.name) > 50) throw new UnprocessableEntityException('The project name is too long')
     if (!_isArray(projectDTO.origins)) throw new UnprocessableEntityException('The list of allowed origins has to be an array of strings')
     if (_size(_join(projectDTO.origins, ',')) > 300) throw new UnprocessableEntityException('The list of allowed origins has to be smaller than 300 symbols')
+    if (_size(_join(projectDTO.ipBlacklist, ',')) > 300) throw new UnprocessableEntityException('The list of allowed blacklisted IP addresses must be less than 300 characters.')
+
+    projectDTO.ipBlacklist.map((ip) => {
```
Comment (kruzhambus) [inline src/project/project.service.ts:224]: using library loadash _map(projectDTO.ipBlacklist, ()=> {})

**Review by kruzhambus — APPROVED** [review]

**Comment by Blaumaus** [comment]
@kruzhambus @rhaxma маски для IP адресов норм будут работать?

**Comment by kruzhambus** [comment]
> @kruzhambus @rhaxma маски для IP адресов норм будут работать?

зараз чекну

**Comment by rhaxma** [comment]
> @kruzhambus @rhaxma маски для IP адресов норм будут работать?

Да.

**Review by kruzhambus — APPROVED** [review]

**Review by Blaumaus — APPROVED** [review]

**Pull request merged**